draft-irtf-cfrg-bbs-signatures.md

```diff
@@ -765,38 +764,104 @@ When defining a new `create_generators` procedure, the most important property i
+
+```
+msg_scalar = messages_to_scalars(messages)
+
+Inputs:
+
+- messages (REQUIRED), a vector of input_messages.
+
+Parameters:
+
+- map_to_scalar, an operation that maps an input_message and its index
+                 to a scalar value, defined by the ciphersuite.
+
+Outputs:
+
+- msg_scalars, a list of scalars.
+
+ABORT if:
+
+1. length(messages) > 2^64 - 1
+
+Procedure:
+
+1. L =  length(messages)
+2. for i in (1, ..., L):
+3.     msg_scalar_i = map_to_scalar(messages[i], i)
+4. return (msg_scalar_1, ..., msg_scalar_L)
+```
+
+### Map to Scalar
+
+As defined above, the `messages_to_scalars` operation works by repeatedly calling the `map_to_scalar` operation, that will be defined by the ciphersuite. The `map_to_scalar` operation accepts the following inputs,
+
+- message (REQUIRED), an input_message that can be either a scalar or an octet string (see [Terminology](#terminology)).
+- index (OPTIONAL), a positive integer. The index the message has in the list of signed messages.
+
+The signature of the operation is the following,
 
-There are multiple ways in which messages can be mapped to their respective scalar values, which is their required form to be used with the [Sign](#signature-generation-sign), [Verify](#signature-verification-verify), [ProofGen](#proof-generation-proofgen) and [ProofVerify](#proof-verification-proofverify) operations.
+```
+msg_scalar = map_to_scalar(msg, index)
+```
+
+Every `map_to_scalar` operation MUST define a unique `MAP_TO_SCALAR_ID` value to be used by the ciphersuite. This value MUST only contain ASCII encoded characters with codes between 0x21 and 0x7e (inclusive) and MUST end with an underscore (ASCII code: 0x5f), other than the last character the string MUST not contain any other underscores (ASCII code: 0x5f).
 
-### Message to Scalar as Hash
+#### Map to Scalar as Hash
 
-This operation takes an input message and maps it to a scalar value via a cryptographic hash function for the given curve. The operation takes also as an optional input a domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuite\_id || "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_", where ciphersuite\_id is the ASCII string representing the unique ID of the ciphersuite "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_" is an ASCII string comprised of 26 bytes.
+This document specifies the following `map_to_scalar` operation, called `map_to_scalar_as_hash`, that uses `hash_to_scalar` as defined in (#hash-to-scalar). Although for extendability reasons, the `map_to_scalar` operation accepts messages that can be either an octet string or a scalar value (as to support protocol specific preprocessing of a message), the `map_to_scalar_as_hash` operation used by this document only maps octet string to scalars and will abort if it gets an `input_message` that is already a scalar value. Additionally, the resulting scalar does not depend on the `index` of the message.
```

Comment:
Hm, this seems to mean that any application which wants to pre-process the messages as mentioned above, must use a ciphersuite not defined by this document, with a different mapping function.

Reply:
Yeap that was the indent. It seems that there is no way to define protocol-agnostic handling of scalar values (i.e., something like `if msg a scalar, return msg` does not work). 

Rules need to be more strict and also consider the index of a message (for example the 1st message is expected to be a scalar, etc). So a ciphersuite defined protocol-specific mapping seems unavoidable.

Reply:
You can always also define the `map_to_scalar` operation to just accept scalars and be the identity, and then apply whatever pre processing you want??